File: 백준/Silver/3976. 역습/역습.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int C = Integer.parseInt(br.readLine());

        for (int t = 0; t < C; t++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int N = Integer.parseInt(st.nextToken());
            int L1 = Integer.parseInt(st.nextToken());
            int L2 = Integer.parseInt(st.nextToken());
            int S1 = Integer.parseInt(st.nextToken());
            int S2 = Integer.parseInt(st.nextToken());

            int[] dp1 = new int[N];
            int[] dp2 = new int[N];
            dp1[0] = L1;
            dp2[0] = L2;

            int[] S1Pass = readArray(br, N - 1);
            int[] S1Drib = readArray(br, N - 1);
            int[] S2Pass = readArray(br, N - 1);
            int[] S2Drib = readArray(br, N - 1);

            for (int j = 1; j < N; j++) {
                dp1[j] = Math.min(dp1[j - 1] + S1Drib[j - 1], dp2[j - 1] + S2Pass[j - 1]);
                dp2[j] = Math.min(dp2[j - 1] + S2Drib[j - 1], dp1[j - 1] + S1Pass[j - 1]);
            }

            int result = Math.min(dp1[N - 1] + S1, dp2[N - 1] + S2);
            System.out.println(result);
        }
    }

    private static int[] readArray(BufferedReader br, int size) throws IOException {
        int[] arr = new int[size];
        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < size; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }
        return arr;
    }
}

File: 백준/Silver/3976. 역습/역습.py
# 역습
import sys
input = sys.stdin.readline

C = int(input())

for i in range(C):
    N, L1, L2, S1, S2 = map(int,input().split())
    dp1 = [0] * 2
    dp2 = [0] * 2
    # 처음 값은 수비수로 부터 패스받은 L1, L2 
    dp1[0] = L1
    dp2[0] = L2

    # 1번 스트라이커의 패스와 드리블 난이도 
    S1Pass = list(map(int,input().split()))
    S1Drib = list(map(int,input().split()))
    # 2번 스트라이커의 패스와 드리블 난이도
    S2Pass = list(map(int,input().split()))
    S2Drib = list(map(int,input().split()))

    # 각 지점의 최소난이도는 드리블을 하거나 패스받은 경우 중 최소값
    for j in range(N-1):
        dp1[1] = min(dp1[0] + S1Drib[j], dp2[0] + S2Pass[j])
        dp2[1] = min(dp2[0] + S2Drib[j], dp1[0] + S1Pass[j])

        dp1[0] = dp1[1]
        dp2[0] = dp2[1]
        
    # 마지막에 슛을 하여 최소값을 출력
    print(min(dp1[0] + S1, dp2[0] + S2))
        
